components/prism-sas.js

```diff
@@ -1,3 +1,29 @@
+(function (Prism) {
+
+	numberPattern = /\b(?:\d[\da-f]*x|\d+(?:\.\d+)?(?:e[+-]?\d+)?)\b/i;
+	numericConstant = {
+	    pattern: /(["'])(?:\1\1|\\.|(!?\1)[^\\])\1[bx]/,
+	    alias: 'number'
+	};
+
+	step = {
+		pattern: /(?:proc\s+\w+|quit|run|data(?!\=))\b/i,
+		alias: 'keyword'
+	};
+
+	comment = [
+		{
+			pattern: /(^\s*|;\s*)\*(.|[\s\S]])*?;/m,
```

Comment:
Typo. There's one `]` too much. But `(.|[\s\S])*?;` itself is an even bigger problem because it fails in exponential time.

I encourage you to try the pattern `(.|[\s\S])*?;` on a site like [regexr.com](https://regexr.com). You'll see that for a string which doesn't contain any `;`s, this pattern fails in O(2^n) (n being the length of the input string). After only 12 characters, this pattern takes >100ms to fail.

I won't go into the finer details but when you have quantified groups, make sure that the alternatives of the group are disjoint (meaning that there is no word which can be matched by more than one alternative of the group). Otherwise, the pattern will fail in O(2^n) on certain input strings.
This is also known under the name [ReDoS](https://en.wikipedia.org/wiki/ReDoS) and is what happened to [Cloudflare](https://blog.securityinnovation.com/cloudflare-dos-blackout) not too long ago although there the cause for the exponential behavior was slightly different.

In this case, `.` is a subset of `[\s\S]` which is both the problem and the reason why it's easy to fix. We can just make it `[\s\S]*?;` and it will still behave the same but it will fail in linear time.
But we can go even simpler because we don't even need the lazy matching.

```suggestion
			pattern: /(^\s*|;\s*)\*[^;]*;/m,
```

Reply:
Thanks, I was not familiar with this site. WIll use it now instead of Regex101.